runtime/src/main/java/org/corfudb/runtime/view/ManagementView.java

```diff
@@ -320,125 +238,303 @@ private NodeStatus getLogUnitNodeStatusInLayout(Layout layout, String server) {
+        } catch (Exception e) {
+            log.error("getClusterStatus[{}]: cluster status unavailable. Exception: {}",
+                    this.runtime.getParameters().getClientId(), e);
+            return getUnavailableClusterStatusReport(layoutServers);
         }
-        return nodeStatusMap;
     }
 
-    /**
-     * Gets the cluster status by pinging each node.
-     * These pings are then compared to the layout to decide whether the cluster is:
-     * STABLE, DEGRADED OR UNAVAILABLE.
-     * The report consists of the following:
-     * Layout - at which the report was generated.
-     * ClientServerConnectivityStatusMap - the connectivity status of this client to the cluster.
-     * ClusterStatus - Health of the Corfu cluster.
-     *
-     * @return ClusterStatusReport
-     */
-    public ClusterStatusReport getClusterStatus() {
+    private ClusterStatusReport getUnavailableClusterStatusReport(List<String> layoutServers, ClusterStatusReliability reliability, @Nonnull Set<String> responsiveServers, Layout layout) {
+        return new ClusterStatusReport(layout,
+                ClusterStatus.UNAVAILABLE,
+                reliability,
+                layoutServers.stream().collect(Collectors.toMap(
+                        endpoint -> endpoint,
+                        node -> NodeStatus.NA)),
+                layoutServers.stream().collect(Collectors.toMap(
+                        endpoint -> endpoint,
+                        endpoint -> responsiveServers.contains(endpoint) ? ConnectivityStatus.RESPONSIVE : ConnectivityStatus.UNRESPONSIVE)));
+    }
+
+    private ClusterStatusReport getUnavailableClusterStatusReport(List<String> layoutServers) {
+        return getUnavailableClusterStatusReport(layoutServers, ClusterStatusReliability.UNAVAILABLE, Collections.emptySet(), null);
+    }
 
-        List<String> layoutServers
-                = new ArrayList<>(runtime.getLayoutServers());
-        Layout layout = getHighestEpochLayout(layoutServers);
-
-        // If layout is null, none of the existing layout servers have a layout.
-        if (layout == null) {
-            return new ClusterStatusReport(null,
-                    layoutServers.stream().collect(Collectors.toMap(
-                            endpoint -> endpoint,
-                            node -> NodeStatus.DOWN)),
-                    ClusterStatus.UNAVAILABLE);
+    private Map<String, Layout> getLayouts(List<String> layoutServers, Boolean discoverLayoutServers) {
+        Map<String, CompletableFuture<Layout>> layoutFuturesMap = getLayoutFutures(layoutServers);
+        Map<String, Layout> layoutMap = new HashMap<>();
+
+        // Wait on the Completable futures
+        for (Entry<String, CompletableFuture<Layout>> entry : layoutFuturesMap.entrySet()) {
+            try {
+                Layout nodeLayout = entry.getValue().get();
```

Comment:
It takes linear time to connect to all servers, it should be run in parallel

Reply:
all calls are async and were done in the previous step, anyways we need to wait on all responses. Anyways, this part of the code will change when we have the quorum read API

Reply:
then, ok